hoomd/state.py

```diff
@@ -81,87 +215,119 @@ def __init__(self, simulation, snapshot):
-        all MPI ranks and from GPUs is gathered on the root MPI rank's memory.
-        When accessing data in MPI simulations, it is recommended to use a
-        ``if snapshot.communicator.rank == 0:`` conditional to prevent
-        attempting to access data on a non-root rank.
-
-        This property can be set to replace the system state with the given
-        `hoomd.Snapshot` object.  Example use cases in which a simulation's
-        state may be reset from a snapshot include Monte Carlo schemes
-        implemented at the Python script level, where the current snapshot is
-        passed to the Monte Carlo simulation before being passed back after
-        running some Monte Carlo steps.
+        """Simulation snapshot.
 
-        Warning:
-            Using `State.snapshot` multiple times will gather data across MPI
-            ranks and GPUs every time. If the snapshot is needed for more than
-            one use, it is recommended to store it in a variable.
+        .. deprecated:: 3.0.0-beta.8
+            Use `take_snapshot` and `restore_snapshot` instead.
+        """
+        warnings.warn("Deprecated, use state.take_snapshot()",
+                      DeprecationWarning)
+        return self.take_snapshot()
+
+    @snapshot.setter
+    def snapshot(self, snapshot):
+        warnings.warn("Deprecated, use state.restore_snapshot()",
+                      DeprecationWarning)
+        self.restore_snapshot(snapshot)
+
+    def take_snapshot(self):
+        """Make a copy of the simulation current state.
+
+        `State.take_snapshot` makes a copy of the simulation state and
+        makes it available in a single object. `State.restore_snapshot` resets
+        the internal state to that in the given snapshot. Use these methods
+        to implement techniques like hybrid MD/MC or umbrella sampling where
+        entire system configurations need to be reset to a previous one after a
+        rejected move.
+
+        Note:
+            Data across all MPI ranks and from GPUs is gathered on the root MPI
+            rank's memory. When accessing data in MPI simulations, use a ``if
+            snapshot.communicator.rank == 0:`` conditional to access data arrays
+            only on the root rank.
 
         Note:
-            Setting or getting a snapshot is an order :math:`O(N_{particles}
-            + N_{bonds} + \ldots)` operation.
+            `State.take_snapshot` is an order :math:`O(N_{particles} + N_{bonds}
+            + \\ldots)` operation.
+
+        See Also:
+            `restore_snapshot`
+
+        Returns:
+            hoomd.Snapshot: The current simulation state
         """
         cpp_snapshot = self._cpp_sys_def.takeSnapshot_double()
         return Snapshot._from_cpp_snapshot(cpp_snapshot,
                                            self._simulation.device.communicator)
 
-    @snapshot.setter
-    def snapshot(self, snapshot):
+    def restore_snapshot(self, snapshot):
```

Comment:
I know `restore_snapshot` is the name used in v2, so there is benefit in using this for v3. However, I don't know if it is the best name. The naming seems to imply the primary use case for the method is to restore the state from a previous snapshot which if that is what we are good for is fine, but if this is meant to be a general purpose setting `set_snapshot`, `apply_snapshot`, or something similar might be a better name.

If you think that continuity here makes sense over any other reasons, then I am fine as is. The name isn't bad; I just wanted to mention this to discuss.

Reply:
Do you think it is OK to have a `set_snapshot` that can only set snapshots with identical particle/bond/etc.. types? I don't have a strong preference. If we choose `set_snapshot` then the corresponding operation should be `get_snapshot`.

Reply:
I think the limitations on behavior are fine, regardless of the naming of `restore/take` or `set/get`. It would be helpful to add a link to the `State.create_state_from_snapshot` method in these setter/getter methods.